### ISSUE TITLE
Proposal: Self-referential functions and the design of options

### ISSUE THREAD

**Issue by puellanivis** [pkg/sftp#189]
Reference: https://commandcenter.blogspot.de/2014/01/self-referential-functions-and-design.html

**Comment by eikenb** [comment]
Could you write a short proposal for specifically what you are proposing?

**Comment by puellanivis** [comment]
```
type Option func(*Client) (Option, error)

func MaxPacket(size int) Option {
	return func(c *Client) (Option, error) {
		if size < 1<<15 {
			return nil, errors.Errorf("size must be greater or equal to 32k")
		}
		save := c.MaxPacket
		c.maxPacket = size
		return MaxPacket(save), nil
	}
}
```

An additional benefit here is that since the options are now a type, all of the functions returning an Option get clustered in the documentation under the Option type, making a quick lookup for humans looking for what options are available. (If there are a lot of types, naming it ClientOption even clusters the documentation for the type near the type to which it applies.)

**Comment by eikenb** [comment]
I see the benefit of the type to cluster the options together. So that seems like a good addition. I'm not sure I see the need for the self-referential bit for this library, particularly given the options are currently passed vie that factory method and thus you wouldn't use them to change values after the fact as  the examples in the blog.

TLDR; I'll add the ClientOption type that fits the options as they are currently implemented.

**Comment by puellanivis** [comment]
Yeah, the self-referential bit is more of a “if you do this from the start, then it’s there, and you can design with it in mind.”

But 90% of the time, I end up throwing them away also, because it doesn’t make any other sense in the constructor to do anything else…

OBTW off-topic: internally at Google calling the New…() functions “factories” was the quickest way I found to getting berated at for not understanding Go… :( “but they’re factories” their response: “we don’t use that word.”

**Comment by eikenb** [comment]
lol... yeah. I think the blessed word is 'constructors'. I'll try to use that in the future so as not to trigger any Googlers. ;)
